src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,261 @@
+  {
+    public:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    public:
+      mmm_module (const char *name, const MMM_COMP_INFO *info);
+      mmm_module () {} /* for dummy */
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+      const char *get_name ();
+      void add_stat (uint64_t size, bool init);
+      void sub_stat (uint64_t size, bool init);
+      void add_expand_count ();
+      const MMM_MEM_STAT &get_stat ();
+      int add_component (const char *name);
+      const std::vector<mmm_component *> &get_comp_vec ();
+      const std::vector<int> &get_comp_idx_map ();
+      void add_comp_idx_map (int compound);
+
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      virtual ~mmm_module ()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+    private:
+      char *m_module_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_component *> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  /* you have to add your modules in Memory Monitoring Manager(MMM) class */
+  class memory_monitoring_manager
+  {
+    public:
+      memory_monitoring_manager (const char *name) : m_aggregater { this }
+      {
+	this->m_server_name = new char[strlen (name) + 1];
+	strcpy (this->m_server_name, name);
+	//m_aggregater = new mmm_aggregater (this);
+      }
+
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+      int sub_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID src, MMM_STAT_ID dest, uint64_t size);
+      int aggregate_module_info (MEMMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MEMMON_TRAN_INFO *info, int tran_count);
+
+      static int comp (const void *a, const void *b)
+      {
+	uint64_t x1 = ((std::pair<int, uint64_t> *)a)->second;
+	uint64_t x2 = ((std::pair<int, uint64_t> *)b)->second;
+
+	if (x1 < x2)
+	  {
+	    return 1;
+	  }
+	else if (x1 > x2)
+	  {
+	    return -1;
+	  }
+
+	return 0;
+      }
+      static inline int MMM_COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return (comp_idx_map_val >> 16);
+      }
+
+      static inline int MMM_SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return (comp_idx_map_val & MMM_PARSE_MASK);
+      }
+
+      ~memory_monitoring_manager ()
+      {
+	delete[] m_server_name;
+	for (int i = 0; i < MMM_MODULE_LAST; i++)
+	  {
+	    delete m_module[i];
+	  }
+      }
+
+    private:
+      class mmm_aggregater
```

Comment:
While I am reviewing, I think we need to make a rules for the naming to save un-necessary communication cost about the naming. For example, some classes have a prefix mmm_ but, others have not. Unforturnately, we don't have a rules for that. How about make a base rules for the naming of the mmm? I request you make a pattern about the following things:
- class/struct/enum
  - member variable, member function
- macro
- fuction
- prefix_ (not only class/function but also others)
  - client, server, mmm etc
- etc - the things you need

And share your naming rules and review if it is needed.

Reply:
I finally decided and shared the naming rule a few days ago.